### DISCUSSION TITLE
Lesson 4: Deploying via Infura

### DISCUSSION THREAD

**Discussion by semsion** [smartcontractkit/full-blockchain-solidity-course-py#268] [Q&A] [answered]
Having tested everything as working fine locally using Ganache, I then went on to test on the Rinkeby test network via Infura (from 4h23m2s https://youtu.be/M576WGiDBdQ?t=15782)

The problem is every time I've run the deploy.py script I get some kind of exception, apart from once, and on this occasion the amount stored via the 'store' function call didn't work, and the returned value was zero (this was returning as 15 locally).

Some of the exceptions I've been getting:

- Could not decode contract function call to retrieve with return data: b'', output_types: ['uint256']
- ValueError: {'code': -32000, 'message': 'replacement transaction underpriced'}
- web3.exceptions.TimeExhausted: Transaction HexBytes('0x9c052b9c907e9540ef064a074a1c61a625dae69e888b58ca9b609b799f8a0fee') is not in the chain, after 120 seconds

Can anyone please clarify why these exceptions and incorrect function call returns might be happening? 

**UPDATE**:

Interestingly, earlier on in the video Patrick added a ‘returns (uint256)’, and ‘return favoriteNumber’ to the SimpleStorage.sol contract. I removed these (as per the version in github), and deploy.py ran as expected for the first time.

Subsequent runs started rendering similar timeout errors though, and also came back with a new error:  

- web3.exceptions.BadFunctionCallOutput: Could not transact with/call contract function, is contract deployed correctly and chain synced?

Any help gratefully received...

**Comment by cromewar** [comment, marked as answer]
Hello @semsion I will try to solve your question:

Could not decode contract function call to retrieve with return data: b'', output_types: ['uint256']:

Make sure the function `retrieve` on your `SimpleStorage.sol` has a return value `uint256`:

``` Solidity
function retrieve() public view returns (uint256){
        return favoriteNumber;
    }
```

Also in the `store` function:

  ```Solidity
  function store(uint256 _favoriteNumber) public {
        favoriteNumber = _favoriteNumber;
    }
```


ValueError: {'code': -32000, 'message': 'replacement transaction underpriced'}

First of all make sure you have enough ETH on your Rinkeby wallet, you can get it from the new faucet provided by chainlink: https://faucets.chain.link/rinkeby

Well the problem occurs because web3 thinks you want to replace an existing uncompleted transaction and it could be solved in different ways.

Easiest one: just delete the compiled_code.json and compile again.
Set the nonce higher: this will say web3 that you are doing a new transaction so:

`nonce = w3.eth.getTransactionCount(my_address) + 20` for example

Increase the gas price: this one is more complex to do and is not worth.

web3.exceptions.TimeExhausted: Transaction HexBytes('0x9c052b9c907e9540ef064a074a1c61a625dae69e888b58ca9b609b799f8a0fee') is not in the chain, after 120 seconds

this is related to the last problem, as the transaction is never completed the default timeout raises the exception.

**Reply by semsion** [reply]
Thanks for the response @cromewar 

My .sol file was as you suggested.

Do have enough ETH.

Have set the nonce to increase by 20.

Am still getting some timeout errors.

Can you clarify which build folder you're referring to?

**Reply by cromewar** [reply]
My bad, build folder is linked to brownie as you are using web3 I meant compiled_code.json, could you please copy-paste your code here please? and also a screenshot of the problem please.

**Reply by semsion** [reply]
As in the compiled_code.json, the python, or the solidity?

**Reply by cromewar** [reply]
I'm specially interested on see your deploy script, please paste it here, don't forget to use
![image](https://user-images.githubusercontent.com/20199180/139090654-1fdc0c95-b637-4779-8441-9f6df929d05a.png)
On github bar to get your code formatted

**Reply by cromewar** [reply]
Also did you update the Keys (private and public) for rinkeby  and added to your env variables?

**Reply by semsion** [reply]
My deploy.py is now exactly the same as the linked below from the course github repo, apart from 'w3', 'my_address', and 'private_key' which are all specific to my details.

I hard coded my private key for now as understand we're using a better method in the brownie lesson.

Public address I took from the Infura Rinkeby test project, and used my MetaMask private key.

https://github.com/PatrickAlphaC/web3_py_simple_storage/blob/main/deploy.py 

I did notice that there's now rinkeby and arbitrum rinkeby endpoints in the Infura project set up. I'm not sure if this would now be causing issues. I'm using the former.

Do you think these issues will not be a problem in the next lesson?

**Reply by cromewar** [reply]
Make sure your w3  host is linked to infura: 

https://rinkeby.infura.io/v3/f6b3restoftheproject

Note: Remember don't share that private key with anyone

if you copy pasted the code from github it should work now, if not... please paste a screenshot of the complete error here.

**Reply by cromewar** [reply]
Yeah there are more endpoint but is just depending on your preference and what blockchain you are working on (we are not using arbitrum on the course)
Also I'm pretty sure you wont have this issue again on the next lesson as your are going to work with brownie which manages all the web3 stuff for you. However I don't like to left problems unsolved so if you still have the same problems you did, I can try to figure out whats happening. 

Best regards

**Reply by semsion** [reply]
@cromewar I very much appreciate your time here.

Yes the w3 host is as we would expect.

The output from the CLI:

``` 
$ ➜  web3_py_simple_storage python deploy.py                  
Installing...
Deploying Contract!
Waiting for transaction to finish...
Done! Contract deployed to <deployed_address>
Initial Stored Value 0
Updating stored Value...
Traceback (most recent call last):
  File "<my_home_path>/.local/lib/python3.8/site-packages/web3/eth.py", line 650, in wait_for_transaction_receipt
    return wait_for_transaction_receipt(self.web3, transaction_hash, timeout, poll_latency)
  File "<my_home_path>/.local/lib/python3.8/site-packages/web3/_utils/transactions.py", line 132, in wait_for_transaction_receipt
    _timeout.sleep(poll_latency)
  File "<my_home_path>/.local/lib/python3.8/site-packages/web3/_utils/threads.py", line 89, in sleep
    self.check()
  File "<my_home_path>/.local/lib/python3.8/site-packages/web3/_utils/threads.py", line 82, in check
    raise self
web3._utils.threads.Timeout: 120 seconds

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "deploy.py", line 89, in <module>
    tx_receipt = w3.eth.wait_for_transaction_receipt(tx_greeting_hash)
  File "<my_home_path>/.local/lib/python3.8/site-packages/web3/eth.py", line 652, in wait_for_transaction_receipt
    raise TimeExhausted(
web3.exceptions.TimeExhausted: Transaction HexBytes('0x0cba17f3d8d737c5214002ac40e9e85c8477d69b752d89f6547358f0a3103dc0') is not in the chain, after 120 seconds
```

**Reply by cromewar** [reply]
Interesting enough I remember I had the same problem when I was on that part of the course (can't remember how I solved it)
Want to try an experiment? 
Use kovan network instead of rinkeby.

1. change to kovan on infura.
2. set the chain id to 42.

![image](https://user-images.githubusercontent.com/20199180/139101858-2e8fd116-8150-4c18-a5b9-fa8bf3926564.png)

3. and of course get some eth for kovan using chainlink faucet: https://faucets.chain.link/kovan

Then try again.

Note: I think the problem is not on your code but on some part of web3 library, anyways double check if your `.sol` file is correct.

**Reply by semsion** [reply]
Hey @cromewar - I'm still getting timeout errors on the Koven testnet. 

I've checked my internet connection and looks good. No obvious packet loss from here...

**Reply by cromewar** [reply]
This is tricky, I've been wondering if hs something to be with your python configuration iteself, could yo create a new python environment (with virutalenv) and run the project there?

Additionally just to verify is nothing wrong on your code, please clone the lesson repo and execute it... does the problem persists?

**Reply by semsion** [reply]
You mean doing something like this: https://www.digitalocean.com/community/tutorials/how-to-install-python-3-and-set-up-a-programming-environment-on-an-ubuntu-20-04-server ??

I did already actually have one of these set up in a different location on this machine.

**Reply by cromewar** [reply]
Yes that is exactly what I meant but it seems you already did that.

This is one of the most weird problems I've ever seen, specially when the code seems to be ok and if you ask me I would bet is related to web3 library, however I don't like to left problems unsolved so I will recreate the code to the same point and try to do more research. Meanwhile I strongly suggest you to continue with the next section of the course, as I said you'll be working with brownie framework which deals with web3 by itself.

Lastly, which OS are you using? this info is crucial to know in which one I should reproduce the project... and I'm sorry I didn't provide you a better solution.

**Reply by semsion** [reply]
Well thank you very much for your effort!

My os is Ubuntu 20.04.3 LTS 
VSCode 1.61.2

**Reply by semsion** [reply]
I was getting an error inside vscode stating that brownie could not be imported. I than ran ‘pip3 list’ to see what was  installed, and could see that brownie wasn't, despite following the instructions. I ran ‘pip3 install eth-brownie’ and vscode could then see the module. 

I was under the impression using pip would use pip3, but it appears not. Maybe this has caused other issues elsewhere.

**Reply by cromewar** [reply]
Interesting enough some windows users had the same problems installing brownie with pipx, as vscode was incapable of recognize brownie.

-  make sure you have selected the correct python interpreter on vscode.
- try installing brownie with pipx
- restart vscode may help

**Reply by semsion** [reply]
I went back and tried the web3_py_simple_storage project again after uninstall (pip and pip3) and reinstall (pip3) of web3, py-solc-x, etc. Still getting a timeout error. 

I'm moving on with the brownie_simple_storage section now.

As I say the brownie import appears fine after installing with pip3.

**Reply by cromewar** [reply]
I've been doing research on the timeout you have, I've reproduced the code on PopOs 21.04 which is basically Ubuntu and I wasn't able to reproduce the problem as everything worked fine on the first run.

Let us know if you have more questions with the following sections.

Best regards, Cromewar

**Reply by semsion** [reply]
Hmm, okay - many thanks for testing.

The first run has sometimes worked, with subsequent runs timing out or returning errors.

Anyway, I'll see how I get on going forward...

**Comment by mjh-jhm** [comment]


**Reply by cromewar** [reply]
Nice, does this work for you @semsion ?

**Comment by fantauzzi** [comment]
Got a "ValueError: replacement transaction underpriced" error when running the integration test `test_can_pick_winner` on Rinkeby, deleting the content of the `build` directory fixed it.
